File: js/src/resources/audio-to-video.ts
import type { HttpClient, PollingOptions, RequestOptions, ActionSchema } from '@runapi.ai/core';
import { compactParams, validateParams } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import { contract } from '../contract_gen';
import type {
  AudioToVideoParams,
  AudioToVideoResponse,
  CompletedAudioToVideoResponse,
  TaskCreateResponse,
} from '../types';

const ENDPOINT = '/api/v1/omnihuman/audio_to_video';

/** Generate a talking-head video from a source image and driving audio. */
export class AudioToVideo {
  constructor(private readonly http: HttpClient) {}

  async run(params: AudioToVideoParams, options?: RequestOptions & PollingOptions): Promise<CompletedAudioToVideoResponse> {
    const { id } = await this.create(params, options);
    const response = await pollUntilComplete<AudioToVideoResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
    return response as CompletedAudioToVideoResponse;
  }

  async create(params: AudioToVideoParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    const body = compactParams(params);
    validateParams(contract['audio-to-video'] as ActionSchema, body as Record<string, unknown>);
    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, {
      body,
      ...options,
    });
  }

  async get(id: string, options?: RequestOptions): Promise<AudioToVideoResponse> {
    return this.http.request<AudioToVideoResponse>('GET', `${ENDPOINT}/${id}`, options ?? {});
  }
}

File: ruby/lib/runapi/omnihuman/resources/audio_to_video.rb
# frozen_string_literal: true

module RunApi
  module Omnihuman
    module Resources
      # Generates talking-head videos from a source image and driving audio.
      class AudioToVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/omnihuman/audio_to_video"
        RESPONSE_CLASS = Types::AudioToVideoResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedAudioToVideoResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["audio-to-video"], params)
        end
      end
    end
  end
end
